conn.go

```diff
@@ -256,31 +254,16 @@ func (s *Session) dialWithoutObserver(ctx context.Context, host *HostInfo, cfg *
 		dialer = d
 	}
 
-	addr := host.HostnameAndPort()
-	conn, err := dialer.DialContext(ctx, "tcp", addr)
-	if err != nil {
-		return nil, err
+	var conn net.Conn
+	var err error
+
+	if endpoint := host.Endpoint(); endpoint != nil {
+		conn, err = s.dialWithEndpoint(ctx, host.Endpoint(), dialer)
```

Comment:
Naming here gets wild can we rename it to doDialEndpoint and doDial?

Reply:
done